prometheusbp/promtest/spec.go

```diff
@@ -0,0 +1,97 @@
+package promtest
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	"github.com/reddit/baseplate.go/errorsbp"
+)
+
+var (
+	errPrefix         = errors.New("the prefix is not at the beginning of the metric name")
+	errLength         = errors.New("metric name should have a minimum of 3 parts, like prefix_name_suffix")
+	errCount          = errors.New("wrong metric count for prefix")
+	errPrometheusLint = errors.New("problem with Prometheus GatherAndLint")
+)
+
+// ValidateSpec validates that the Prometheus metrics being exposed conform to
+// the baseplate spec. This test will validate all metrics that begin with the
+// metricPrefix and will validate the number of metrics with that prefix match
+// wantMetricCount.
+func ValidateSpec(t *testing.T, metricPrefix string, wantMetricCount int) {
+	var batchErrs errorsbp.Batch
+	gotMetricCount, errs := validateSpec(metricPrefix)
+	if len(errs) > 0 {
+		batchErrs.Add(errs...)
+	}
+	if gotMetricCount != wantMetricCount {
+		batchErrs.Add(fmt.Errorf("%w: got %d, want %d", errCount, gotMetricCount, wantMetricCount))
+	}
+
+	for _, e := range batchErrs.GetErrors() {
+		t.Errorf("err: %v", e.Error())
+	}
+}
+
+func validateSpec(metricPrefix string) (int, []error) {
+	var metricCount int
+	var batchErrs errorsbp.Batch
+
+	metricFam, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		return metricCount, batchErrs.GetErrors()
+	}
+
+	for _, m := range metricFam {
```

Comment:
I would make a map of the required metrics and delete them as you see them, and then add an error for each missing metric at the end.  Right now you are just checking for a count, which means that it would be easy to get the wrong metrics and not fail the spec test.